compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUDistributeSharedMemcpyV2.cpp

```diff
@@ -0,0 +1,192 @@
+                                            ArrayRef<Range> parallelLoopRanges,
+                                            Value flatThreadId) {
+  SmallVector<linalg::ProcInfo> infos;
+  Value id = flatThreadId;
+  AffineExpr d0 = b.getAffineDimExpr(0);
+  for (Range r : llvm::reverse(parallelLoopRanges)) {
+    linalg::ProcInfo info;
+    auto offset = r.offset.dyn_cast<Attribute>();
+    auto stride = r.stride.dyn_cast<Attribute>();
+    auto size = r.size.dyn_cast<Attribute>();
+    assert(offset && stride && size);
+    int64_t numThreadsDim = (llvm::cast<IntegerAttr>(size).getInt() -
+                             llvm::cast<IntegerAttr>(offset).getInt()) /
+                            llvm::cast<IntegerAttr>(stride).getInt();
+    Value dimId = id;
+    if (infos.size() != parallelLoopRanges.size() - 1)
+      dimId =
+          affine::makeComposedAffineApply(b, loc, d0 % numThreadsDim, {dimId});
+    info.procId = dimId;
+    info.nprocs = b.create<arith::ConstantIndexOp>(loc, numThreadsDim);
+    info.distributionMethod =
+        linalg::DistributionMethod::CyclicNumProcsGeNumIters;
+    infos.push_back(info);
+    id = affine::makeComposedAffineApply(b, loc, d0.floorDiv(numThreadsDim),
+                                         {id});
+  }
+  std::reverse(infos.begin(), infos.end());
+  return infos;
+}
+
+/// Return a flattened Id Value by combining the 3D gpu thread IDs.
+static Value createFlatId(mlir::FunctionOpInterface funcOp,
+                          ArrayRef<int64_t> workgroupSize) {
+  OpBuilder b(funcOp.getFunctionBody());
+  Type indexType = b.getIndexType();
+  AffineExpr d0 = getAffineDimExpr(0, b.getContext());
+  AffineExpr d1 = getAffineDimExpr(1, b.getContext());
+  AffineExpr d2 = getAffineDimExpr(2, b.getContext());
+  Value threadX =
+      b.create<gpu::ThreadIdOp>(funcOp.getLoc(), indexType, gpu::Dimension::x);
+  Value threadY =
+      b.create<gpu::ThreadIdOp>(funcOp.getLoc(), indexType, gpu::Dimension::y);
+  Value threadZ =
+      b.create<gpu::ThreadIdOp>(funcOp.getLoc(), indexType, gpu::Dimension::z);
+  Value flatThreadId = affine::makeComposedAffineApply(
+      b, funcOp.getLoc(),
+      d0 + workgroupSize[0] * d1 + (workgroupSize[0] * workgroupSize[1]) * d2,
+      {threadX, threadY, threadZ});
+  return flatThreadId;
+}
+
+namespace {
+struct LLVMGPUDistributeSharedMemcpyV2Pass
+    : public LLVMGPUDistributeSharedMemcpyV2Base<
+          LLVMGPUDistributeSharedMemcpyV2Pass> {
+  void getDependentDialects(DialectRegistry &registry) const override {}
+  void runOnOperation() override {
+    auto funcOp = getOperation();
+    FailureOr<IREE::HAL::ExecutableExportOp> exportOp = getEntryPoint(funcOp);
+    if (failed(exportOp)) {
+      // We cannot do anything because we do not have the workgroup size
+      // information, but the pass did not fail.
+      return;
+    }
+
+    MLIRContext *ctx = &getContext();
+    auto workgroupSize = getWorkgroupSize(exportOp.value());
```

Comment:
Most of the logics are from https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeSharedMemoryCopy.cpp but it uses `CyclicNumProcsGeNumIters` distribution option.